pkg/webhook/cluster/mutation/mutation.go

```diff
@@ -89,12 +103,29 @@ func (h *AdmissionHandler) Handle(ctx context.Context, req webhook.AdmissionRequ
 			return admission.Errored(http.StatusBadRequest, err)
 		}
 
+		// Prevent unconditional CNI upgrades for old clusters.
+		// Do not default cni if it was not explicitly set.
+		preventCNIDefaulting := cluster.Spec.CNIPlugin == nil
+
+		// apply defaults to the new version
+		err := h.applyDefaults(ctx, cluster)
+		if err != nil {
+			h.log.Info("cluster mutation failed", "error", err)
+			return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("cluster mutation request %s failed: %v", req.UID, err))
+		}
+
+		if preventCNIDefaulting {
+			cluster.Spec.CNIPlugin = nil
+		}
```

Comment:
Why don't we default this to a proper value? Please correct me if I'm mistaken, but don't we reliably know that if `cluster.Spec.CNIPlugin == nil` then the CNI plugin is Canal v3.9? The purpose of defaulting is to populate as many fields as possible so we don't have nils -- this simplifies logic later significantly.

Reply:
There is custom logic for upgrading or setting CNI's. We do not want  this to happen automatically for all old clusters as we cannot ensure that this wont break existing clusters.  See: https://github.com/kubermatic/kubermatic/blob/0ab42f058ca9e693b506569c5737e42daa153142/pkg/validation/cluster.go#L710